lib/src/rewrite.rs

```diff
@@ -710,17 +710,33 @@ pub fn move_commits(
             if let Some(new_child_parents) = new_children_parents.get(commit_id) {
                 new_child_parents.clone()
             }
-            // Commits in the target set should persist only rebased parents from the target
-            // sets.
-            else if let Some(target_commit_parents) =
-                target_commits_internal_parents.get(commit_id)
-            {
-                // If the commit does not have any parents in the target set, it is one of the
-                // commits in the root set, and should be rebased onto the new destination.
-                if target_commit_parents.is_empty() {
+            // Commit is in the target set.
+            else if target_commit_ids.contains(commit_id) {
+                // If the commit is a root of the target set, it should be rebased onto the new destination.
+                if target_roots.contains(commit_id) {
                     new_parent_ids.clone()
-                } else {
-                    target_commit_parents.clone()
+                }
+                // Otherwise:
+                // 1. Keep parents which are within the target set.
+                // 2. Replace parents which are outside the target set but are part of the
+                //    connected target set with their ancestor commits which are in the target
+                //    set.
+                // 3. Keep other parents outside the target set if they are not descendants of the
+                //    new children of the target set.
+                else {
+                    let mut new_parents = vec![];
+                    for parent_id in commit.parent_ids() {
+                        if target_commit_ids.contains(parent_id) {
+                            new_parents.push(parent_id.clone());
+                        } else if let Some(parents) =
+                                connected_target_commits_internal_parents.get(parent_id) {
+                            new_parents.extend(parents.iter().cloned());
+                        } else if !new_children.iter().any(|new_child| {
+                                mut_repo.index().is_ancestor(new_child.id(), parent_id) }) {
+                            new_parents.push(parent_id.clone());
+                        }
+                    }
```

Comment:
Any thoughts on https://github.com/martinvonz/jj/pull/3605#discussion_r1587018019 ?

I don't think it would matter in practice, but I think it's nice that both internal and external edges are processed somewhat consistently. Only the roots are detached.

This is not a blocker, btw.

Reply:
I will spend some time (probably next week) to look at this.